public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts

```diff
@@ -0,0 +1,151 @@
+      resources 
+      | where resourceGroup == "${resourceGroup.name.toLowerCase()}"
+      | where type in (${SUPPORTED_RESOURCE_TYPES}) and location in (${SUPPORTED_LOCATIONS})
+    `);
+
+    // TODO: figure out desired error handling strategy
+    if (!ok) {
+      throw new Error('unable to fetch resource containers');
+    }
+
+    return this._formatAndCacheResourceData(response.data);
+  }
+
+  _formatAndCacheResourceData(rawData: RawAzureResourceGraphDataItem[]) {
+    // "produce" from immer takes in an object, allows you to edit a copy of that object and then returns a new object
+    this.formattedResourcePickerData = produce(this.formattedResourcePickerData, (draftState: any) => {
+      rawData.forEach((item: RawAzureResourceGraphDataItem) => {
+        switch (item.type) {
+          case 'microsoft.resources/subscriptions':
+            draftState[item.subscriptionId] = {
+              // handles potential duplicates (see note on proccessing resource groups before subscriptions)
+              ...(draftState[item.subscriptionId] ? draftState[item.subscriptionId] : {}),
+
+              name: item.name,
+              id: item.subscriptionId,
+              subscriptionId: item.subscriptionId,
+              typeLabel: 'Subscription',
+              type: EntryType.Collection,
+              children: {},
+            };
+
+            break;
+
+          case 'microsoft.resources/subscriptions/resourcegroups':
+            // handles processing resourcegroup before we encounter it's parent subscription
+            if (!draftState[item.subscriptionId]) {
+              draftState[item.subscriptionId] = {
+                children: {},
+              };
+            }
+
+            // a note here that we store resource groups by name and not by id because the resources we fetch only have resource group names not ids
+            draftState[item.subscriptionId].children[item.name.toLowerCase()] = {
+              name: item.name,
+              id: item.id,
+              subscriptionId: item.subscriptionId,
+              type: EntryType.SubCollection,
+              typeLabel: 'Resource Group',
+              children: {},
+            };
+            break;
+
+          default:
+            // We assume everything else to be a selectable resource
+            draftState[item.subscriptionId].children[item.resourceGroup].children[item.id] = {
+              name: item.name,
+              id: item.id,
+              type: EntryType.Resource,
+              typeLabel: item.type, // TODO: these types can be quite long, we may wish to format them more
+              location: item.location, // TODO: we may wish to format these locations, by default they are written as 'northeurope' rather than a more human readable "North Europe"
+            };
+        }
+      });
+    });
+
+    return this.formattedResourcePickerData;
+  }
+
+  async _makeResourceGraphRequest(query: string, maxRetries = 1): Promise<any> {
```

Comment:
Would be nice if we could type this a little bit cleaner, without the any 😅

It looks like it should be something like `Promise<FetchResponse<{ data: RawAzureResourceGraphDataItem[] }>>` ?

Maybe we could make the response object (` { data: ... } `) its own interface?